File: rehovot-ads-packages/ios/Plugin/UnityLevelPlayPlugin.swift
import Capacitor
import Foundation

// Capacitor bridge only; all native behavior stays behind CapacitorAds.
@objc(CapacitorAdsPlugin)
public class CapacitorAdsPlugin: CAPPlugin {
    private var capacitorAds: CapacitorAds?

    public override func load() {
        capacitorAds = CapacitorAds(logger: Logger(tag: "CapacitorAds"))
    }

    @objc func initialize(_ call: CAPPluginCall) {
        capacitorAds?.initialize(call.toInitializeRequest())
        call.resolve()
    }

    @objc func showBanner(_ call: CAPPluginCall) {
        capacitorAds?.showBanner(call.toBannerRequest())
        call.resolve()
    }

    @objc func hideBanner(_ call: CAPPluginCall) {
        capacitorAds?.hideBanner()
        call.resolve()
    }

    @objc func destroyBanner(_ call: CAPPluginCall) {
        capacitorAds?.destroyBanner()
        call.resolve()
    }

    @objc func showInterstitial(_ call: CAPPluginCall) {
        capacitorAds?.showInterstitial(call.toInterstitialRequest())
        call.resolve()
    }

    @objc func showRewarded(_ call: CAPPluginCall) {
        guard let result = capacitorAds?.showRewarded(call.toRewardedRequest()) else {
            call.resolve([
                "success": false,
                "completed": false,
                "message": "CapacitorAds bridge is not ready.",
            ])
            return
        }

        var response: [String: Any] = [
            "success": result.success,
            "completed": result.completed,
            "message": result.message,
        ]

        if let placementId = result.placementId {
            response["placementId"] = placementId
        }

        if let rewardAmount = result.rewardAmount {
            response["rewardAmount"] = rewardAmount
        }

        if let rewardCurrency = result.rewardCurrency {
            response["rewardCurrency"] = rewardCurrency
        }

        call.resolve(response)
    }

    @objc func destroy(_ call: CAPPluginCall) {
        capacitorAds?.destroy()
        call.resolve()
    }
}

private extension CAPPluginCall {
    func toInitializeRequest() -> InitializeRequest {
        return InitializeRequest(
            appKey: getString("appKey"),
            testMode: getBool("testMode") ?? false,
            debug: getBool("debug") ?? false,
            autoStart: getBool("autoStart") ?? true
        )
    }

    func toBannerRequest() -> BannerRequest {
        return BannerRequest(
            placementId: getString("placementId"),
            position: getString("position")
        )
    }

    func toInterstitialRequest() -> InterstitialRequest {
        return InterstitialRequest(
            placementId: getString("placementId"),
            timeoutMs: getInt("timeoutMs")
        )
    }

    func toRewardedRequest() -> RewardedRequest {
        return RewardedRequest(
            placementId: getString("placementId"),
            rewardAmount: getInt("rewardAmount"),
            rewardCurrency: getString("rewardCurrency")
        )
    }
}

File: rehovot-ads-packages/android/src/main/java/com/rehovottech/levelplay/UnityLevelPlayPlugin.kt
package com.rehovottech.capacitorads

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

// Capacitor bridge only; all native behavior stays behind CapacitorAds.
@CapacitorPlugin(name = "CapacitorAds")
class CapacitorAdsPlugin : Plugin() {
    private lateinit var capacitorAds: CapacitorAds

    override fun load() {
        capacitorAds = CapacitorAds(Logger("CapacitorAds", false))
    }

    @PluginMethod
    public fun initialize(call: PluginCall) {
        capacitorAds.initialize(call.toInitializeRequest())
        call.resolve()
    }

    @PluginMethod
    public fun showBanner(call: PluginCall) {
        capacitorAds.showBanner(call.toBannerRequest())
        call.resolve()
    }

    @PluginMethod
    public fun hideBanner(call: PluginCall) {
        capacitorAds.hideBanner()
        call.resolve()
    }

    @PluginMethod
    public fun destroyBanner(call: PluginCall) {
        capacitorAds.destroyBanner()
        call.resolve()
    }

    @PluginMethod
    public fun showInterstitial(call: PluginCall) {
        capacitorAds.showInterstitial(call.toInterstitialRequest())
        call.resolve()
    }

    @PluginMethod
    public fun showRewarded(call: PluginCall) {
        val result = capacitorAds.showRewarded(call.toRewardedRequest())
        val response = JSObject()
            .put("success", result.success)
            .put("completed", result.completed)
            .put("message", result.message)

        response.put("placementId", result.placementId)
        response.put("rewardAmount", result.rewardAmount)
        response.put("rewardCurrency", result.rewardCurrency)
        call.resolve(response)
    }

    @PluginMethod
    public fun destroy(call: PluginCall) {
        capacitorAds.destroy()
        call.resolve()
    }

    private fun PluginCall.toInitializeRequest(): InitializeRequest {
        return InitializeRequest(
            appKey = getString("appKey"),
            testMode = getBoolean("testMode") == true,
            debug = getBoolean("debug") == true,
            autoStart = getBoolean("autoStart") != false,
        )
    }

    private fun PluginCall.toBannerRequest(): BannerRequest {
        return BannerRequest(
            placementId = getString("placementId"),
            position = getString("position"),
        )
    }

    private fun PluginCall.toInterstitialRequest(): InterstitialRequest {
        return InterstitialRequest(
            placementId = getString("placementId"),
            timeoutMs = getInt("timeoutMs"),
        )
    }

    private fun PluginCall.toRewardedRequest(): RewardedRequest {
        return RewardedRequest(
            placementId = getString("placementId"),
            rewardAmount = getInt("rewardAmount"),
            rewardCurrency = getString("rewardCurrency"),
        )
    }
}
